tensorizer/serialization.py

```diff
@@ -2432,384 +2181,283 @@ def _generate_state_dict(self) -> None:
+        # there are probably algorithms to do this more optimally
+        tensor_sizes = [(name, self._metadata[name].deserialized_length) for name in keys]
+        if self._num_readers == 1:
+            tensors_per_reader = [tensor_sizes]
+        else:
+            if not isinstance(self._file_spec, (str, bytes, os.PathLike, int)):
+                raise Exception("File specifier must be a string to support concurrent readers ")
+
+            total_tensor_bytes = sum(tensor_size for _, tensor_size in tensor_sizes)
+            ideal_tensor_bytes_per_reader = int(total_tensor_bytes / self._num_readers)
+
+            tensors_per_reader = []
+            running_total = 0
+            chunk_start = 0
+            for tensor_size_idx in range(len(tensor_sizes)):
+                # if this tensor would make this current reader more than ideal_tensor_bytes_per_reader, and there's at least one existing item in the reader
+                if running_total + tensor_sizes[tensor_size_idx][1] > ideal_tensor_bytes_per_reader and chunk_start != tensor_size_idx:
+                    # break it
+                    tensors_per_reader.append(tensor_sizes[chunk_start:tensor_size_idx])
+                    chunk_start = tensor_size_idx
+                    running_total = 0
+                running_total += tensor_sizes[tensor_size_idx][1]
+
+            if running_total:
+                # The last one gets the leftovers
+                if not tensors_per_reader:
+                    tensors_per_reader.append([])
+
+                tensors_per_reader[-1].extend(tensor_sizes[chunk_start:])
+
+        transfer_out_queue = queue.SimpleQueue()  # type: queue.SimpleQueue[Union[Exception, TensorDeserializer._CopiedData]]
+
+        copy_threads = []
+        barrier = threading.Barrier(len(tensors_per_reader))
+        halt = AtomicUint(width=4)
+
+        for thread_idx, tensor_items in enumerate(tensors_per_reader):
+            # print(f'Thread #{thread_idx} will read {len(tensor_items)} tensors, {sum(v for _, v in tensor_items)/1024/1024:.3f}MiB, {max(v for _, v in tensor_items)/1024/1024:.3f}MiB max', file=sys.stderr)
+
+            thread = threading.Thread(
+                name=f'TensorDeserializerCopy-{thread_idx}',
+                target=self._copy_thread,
+                args=(
+                    thread_idx,
+                    halt,
+                    barrier,
+                    verify_hash,
+                    tensor_items,
+                    transfer_out_queue)
+            )
+            copy_threads.append(thread)
+            thread.start()
+
+        for _ in range(len(keys)):
+            copied_data = transfer_out_queue.get()
+            if isinstance(copied_data, Exception):
+                halt.store(1)
+                # error occurred; halt
+                raise copied_data
+
+            name = copied_data.header.name
+            self._cache[name] = copied_data
+            yield copied_data
```

Comment:
The ordering guarantee for multiple readers can be restored for compatibility in certain scenarios by using a min-heap weighted by tensor index and popping whenever `heap[0]` is the next in-order tensor instead of yielding immediately. This also allows requesting out-of-order tensors efficiently for a single reader, since they can be sorted and have their original order restored at `yield`-time. The implementation should be fairly quick using `heapq`.

Reply:
I'll add this for now for compatibility, but imho I don't love stalling results so that they are returned in order. I worry the existing interface limits how fast we could be. Happy to measure the impact of this and revisit in a future rev tho.